src/software/simulated_tests/avoid_collisions.py

```diff
@@ -0,0 +1,144 @@
+            for j in range(i + 1, len(robots)):
+                if self.check_robot_collision(robots[i], robots[j]):
+                    return ValidationStatus.FAILING
+
+        return ValidationStatus.PASSING
+
+    def check_robot_collision(self, robot1, robot2):
+        """
+        Helper function to check if 2 robots have collided
+        Also determines which robots have committed a foul
+        :param robot1: first robot
+        :param robot2: second robot
+        :return: True if a collision occurs
+                 False if no collision
+                 Adds robots which fouled to self.fouled_robots
+        """
+
+        # robot positions
+        robot_1_pos = tbots.createVector(
+            Vector(
+                x_component_meters=robot1.current_state.global_position.x_meters,
+                y_component_meters=robot1.current_state.global_position.y_meters,
+            )
+        )
+
+        robot_2_pos = tbots.createVector(
+            Vector(
+                x_component_meters=robot2.current_state.global_position.x_meters,
+                y_component_meters=robot2.current_state.global_position.y_meters,
+            )
+        )
+
+        # robot velocities
+        robot_1_vel = tbots.createVector(
+            Vector(
+                x_component_meters=robot1.current_state.global_velocity.x_component_meters,
+                y_component_meters=robot1.current_state.global_velocity.y_component_meters,
+            )
+        )
+
+        robot_2_vel = tbots.createVector(
+            Vector(
+                x_component_meters=robot2.current_state.global_velocity.x_component_meters,
+                y_component_meters=robot2.current_state.global_velocity.y_component_meters,
+            )
+        )
+
+        # check if robots are colliding
+        if (robot_1_pos - robot_2_pos).length() < ROBOT_MAX_RADIUS_METERS * 2 + 0.005:
+            # check if the projection of the difference between their velocities
+            # onto the line between their positions > 1.5
+            # if this case is true, robots are considered to have collided
+            # but still have to determine which ones have fouled
+            if (robot_1_vel - robot_2_vel).dot(robot_1_pos - robot_2_pos) / (
+                robot_1_pos - robot_2_pos
+            ).length() > 1.5:
+                # checks if the absolute speed difference between the robots is < 0.3
+                # if so, both get a foul
+                if robot_1_vel.length() - robot_2_vel.length() < 0.3:
+                    self.fouled_robots.extend([robot1.id, robot2.id])
+                else:
+                    # checks which robot is faster (that one will get a foul)
+                    if robot_1_vel.length() > robot_2_vel.length():
+                        self.fouled_robots.extend([robot1.id])
+                    else:
+                        self.fouled_robots.extend([robot2.id])
+                return True
```

Comment:
Could you also link the SSL rule where this logic is from: https://robocup-ssl.github.io/ssl-rules/sslrules.html#_crashing